Multi project extension

MysticJay opened the pull request:
This is the first and IMHO most important plugin ZasoGD has created.
The script itself is unchanged compared to the original file.
Headers and footers are changed (not sure about the `@match` line)

Zaso's startup wrapper has been commented out, although I do not yet understand why.

The script has been tested in IITC-CE - Menu item load.

MysticJay committed:
initial import

no changes to Zaso's script.

MysticJay committed:
Obvious changes done

Header, Tail
wrapper () removed

MysticJay committed:
Create Multi-Projects-Extension.doc

Code Snippet for integrating other plugins with MPE

modos189 commented on plugins/multi-projects-extension.user.js:
```diff
@@ -0,0 +1,544 @@
+// @name           IITC plugin: Multi Projects Extension
+// @category       Control
+// @version        0.0.6.@@DATETIMEVERSION@@
+// @namespace      http://www.giacintogarcea.com/ingress/items/
```
Please remove lines @namespace, @updateURL, @downloadURL, @match and @grand

MysticJay committed:
Update multi-projects-extension.user.js

removed commented functions and @match

modos189 commented on plugins/multi-projects-extension.user.js:
```diff
@@ -0,0 +1,544 @@
+@@METAINFO@@
+// ==/UserScript==
+@@PLUGINSTART@@
+// function wrapper(){
```
The wrapper is added when building the project, please remove these comments

modos189 commented on plugins/multi-projects-extension.user.js:
```diff
@@ -0,0 +1,544 @@
+
+@@PLUGINEND@@
+
+// if(window.iitcLoaded && typeof setup === 'function'){
```
and this comments

MysticJay commented:
IMHO ready to merge

modos189 commented on plugins/multi-projects-extension.user.js:
```diff
@@ -0,0 +1,544 @@
+// PLUGIN START ////////////////////////////////////////////////////////
+
+	// use own namespace for plugin
+	window.plugin.mpe = function() {};
```
```suggestion
window.plugin.mpe = function() {};
```

modos189 replied:
Reduce the indent for all code.

Everything else is fine.

modos189 commented:
Contents of `Multi-Projects-Extension.doc` file please place as comment in multi-projects-extension.user.js file, and delete Multi-Projects-Extension.doc

MysticJay commented:
Does IITC-CE work with Block-Comments? /* and */?

MysticJay committed:
Code Snippet moved

Sample Code moved to main script
DOC removed

modos189 commented:
You can use multiline comments. Yeah, like you just did.

We have documentation (in the doc folder) but I don't like it very much, it should be generated based on comments in the code.

modos189 commented:
Pay attention to my two remaining comments https://github.com/IITC-CE/ingress-intel-total-conversion/pull/221/files

MysticJay committed:
Update multi-projects-extension.user.js

Header fixed,
Indentation reduced

modos189 commented:
Thanks, this PR will be merged into a new release

MysticJay committed:
Create font-awesome.user.js

johnd0e commented:
> Thanks, this PR will be merged into a new release

I'd rather think about include support of this in IITC core.
May be in some other form (need to investigate functionality deeper).

MysticJay committed:
Create draw2-beta.user.js

Added MPE-Support for Drawtools.

MysticJay commented:
To test the functionality:
1) import Multi-Project-Extension.user.js to externals
2) import DRAW2-beta.user.js to externals
3) disable regular DRAW-Tools
4) enable MPE and Draw2-beta
Create a drawing
from MultiProjects Menu click on the "+" and create a new project. 
Existing drawing should be cleared.
create a new draw.
Use the dialogue to switch between drawings
Use the settings dialogue to add the selection menu to the MainMenu.

MysticJay committed:
Create bkmkrs2-beta.user.js

Added MPE-Support to current Bookmarks.
import to extrenals and disable standard BMRKS-plugin to test

MysticJay commented:
To test:
disable standard Bookmarks-Plugin,
import bkmks2.beta to externals

johnd0e commented:
1. I do not get why you need separate plugin (draw2). You may just extend single draw-tools (in runtime)

2. But if you still need draw2 then please make your changes reviewable:
   - copy `draw-tools.user.js` to `draw-tools2.user.js`
   - make your changes in separate commit(s)

MysticJay commented:
Bare with me, I am new to GIT, so it seemed to be safer to create a separate draw than to fiddle around in the original draw. At least this way does not break the existing tools and existing extensions to it. When I started porting Zaso's changes it was not clear to me how deep the changes would be.

The changes can easily be followed by just renaming DRAW2-beta.user.js to draw-tools.user.js
Also the contributions should show the changes step by step, as the first contribution of the file was the original draw-tools (just a copy)

Draw2 and Bkmarks2 is "only" a Draft to show how MPE is working. Due to the same namespace both use with the old Versions, the data is completely compatible. If you find a problem with MPE it is easy to switch back to the old plugins without loosing anything.

Also there are many other proposals that need to be reviewed and I think it is easier to add these to a forked plugin and finally just replace the old plugin than to daily send in new contributions.

As you wrote earlier we should rather think of adding MPE to the code as "Projects" or "Project-Manager".

johnd0e commented:
> Also the contributions should show the changes step by step, as the first contribution of the file was the original draw-tools (just a copy)

Yeah, but for `draw-tools` I see only single commit https://github.com/IITC-CE/ingress-intel-total-conversion/pull/221/commits/9d89089e8f3b14fada5be8e1e25543940d69297a

Better if there were (1) initial (2) changes for MPE

> Also there are many other proposals that need to be reviewed and I think it is easier to add these to a forked plugin

We can play with forked plugins, but I do not think that we should release them as they are. It'd be confusing.

> As you wrote earlier we should rather think of adding MPE to the code as "Projects" or "Project-Manager".

Yes. It is general principle.
But there may be different implementations of Multi-Projects idea.
I need time to check it.

MysticJay committed:
Multi-Project-Support

Multi-Project support for Bookmarks and Draw-Tools added.
(imported changes from beta)

MysticJay commented:
Changes applied on original files. removed development (beta) files.

johnd0e commented:
Ok. I will review proposed changes ASAP (but do not expect it to be fast, definitely not earlier than 0.30 release).

johnd0e commented:
Some general thoughts:

MPE allows to group Draws/Bookmarks by related project.
What if we need some particular set of Draws or Bookmarks in several projects?
Oops. We can't.

I suppose we could improve project idea.
E.g.:
Bookmarks already has folders. So we might just select existent folders to add them into project.

johnd0e commented on plugins/font-awesome.user.js:
```diff
@@ -0,0 +1,30 @@
+// ==UserScript==
+// @id             iitc-plugin-font-awesome@Zaso
+// @name           IITC plugin: Font Awesome
```
If we include web font - we should include it in core, to make it available out-of-the-box.
(and it is still unclear if should include - #236)

johnd0e commented on plugins/font-awesome.user.js:
```diff
@@ -0,0 +1,30 @@
+	$('<link>')
+		.prop('type', 'text/css')
+		.prop('rel', 'stylesheet')
+		.prop('href', 'https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css')
```
Why use 4.7 instead of current (5.10.1)?
https://fontawesome.com/

MysticJay replied:
No objection to use 5.10.1.
Just not tested it.

MysticJay commented:
MPE is about replacing the complete namespace of a plugin. Sharing parts between different namespaces would require them to be in an additional namespace.
Whatever I do not think this is needed Copy paste between different projects is possible and helps to avoid accidental sharing of confidential Information between projects.
I highly recommend to not share anything between projects.

johnd0e commented:
I still don't get what do you mean. Explain please.

MysticJay commented:
You asked about sharing parts of draw or bookmarks between projects.

MysticJay commented:
What about moving this to the core?

johnd0e commented:
It is planned

MysticJay committed:
Update multi-projects-extension.user.js

Version alinged with ZasoItems-CE
